library/show.lisp

```diff
@@ -0,0 +1,214 @@
+(coalton-library/utils:defstdlib-package #:coalton-library/show
+  (:documentation "A way to show linear, human-readable, textual representations of objects.")
+  (:use
+   #:coalton
+   #:coalton-library/builtin
+   #:coalton-library/functions
+   #:coalton-library/classes)
+  (:export
+   #:LispOutputStream
+   #:standard-output
+   #:error-output
+   #:finish-output
+   #:make-string-output
+   #:with-show-stream
+
+   #:Show
+   #:show-to
+   #:show-to-string
+   #:newline
+   #:show*
+
+   #:Reveal
+   #:Expose
+   ))
+
+(in-package #:coalton-library/show)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+(cl:defvar *show-stream* (cl:make-synonym-stream 'cl:*standard-output*)
+  "The stream to which `show` will show to by default.")
+
+(coalton-toplevel
+  (repr :native cl:stream)
+  (define-type LispOutputStream
+    "A Lisp output stream.")
+
+  (inline)
+  (declare standard-output (Unit -> LispOutputStream))
+  (define (standard-output)
+    "Lisp's `*standard-output*`. This value is dynamic."
+    (lisp LispOutputStream ()
+      cl:*standard-output*))
+
+  (inline)
+  (declare error-output (Unit -> LispOutputStream))
+  (define (error-output)
+    "Lisp's `*error-output*`. This value is dynamic."
+    (lisp LispOutputStream ()
+      cl:*error-output*))
+
+  (declare finish-output (LispOutputStream -> Unit))
+  (define (finish-output s)
+    "Finish all output on the stream `s`."
+    (lisp Unit (s)
+      (cl:finish-output s)
+      Unit))
+
+  (declare make-string-output (Unit -> (Tuple LispOutputStream (Unit -> String))))
+  (define (make-string-output)
+    "Make a string output stream. Return a tuple of two values:
+
+1. The `LispOutputStream` that can be written to.
+
+2. A thunk which extracts (and clears) the characters written to the stream."
+    (let ((stream (lisp LispOutputStream ()
+                    (cl:make-string-output-stream)))
+          (extract (fn ()
+                     (lisp String (stream)
+                       (cl:get-output-stream-string stream)))))
+      (Tuple stream extract)))
+
+  (inline)
+  (declare show-stream (Unit -> LispOutputStream))
+  (define (show-stream)
+    (lisp LispOutputStream ()
+      *show-stream*))
+
+  (declare with-show-stream (LispOutputStream -> (Unit -> :a) -> :a))
+  (define (with-show-stream s f)
+    "Call the thunk `f`, overriding as `s` the stream as that which `show`
+will write to."
+    (lisp :a (s f)
+      (cl:let ((*show-stream* s))
+        (call-coalton-function f Unit)))))
+
+(coalton-toplevel
+  (define-class (Show :a)
+    "Objects which have a convenient, textual, linear printed representation for
+display in a terminal. This is principally for program output and debugging."
+    (show-to
+     "Display an object to a Lisp stream."
+     (LispOutputStream -> :a -> Unit)))
```

Comment:
Maybe I'm missing something obvious, but couldn't we just do it the way Haskell does it and do the simpler:
```lisp
(define-class (Show :a)
    (show (:a -> String)))
```
Then we could have functions keyed off that which output to streams, etc.

I'm not saying we should always just do things the same way Haskell does them. But in this case what is the benefit of the complexity? And it's not just Haskell that does this. Every other language besides CL I can think of has its override-able "convert this object into a string" function morally equivalent to (Mytype -> String).

Reply:
One possible concern is that Coalton is eager, so when you `show` huge structure you create all the temporary strings before they are output?

Reply:
Gauche has rope-like library that allows O(1) string concatenation.  It is rudimental and trivial, so any library author can hack up its own, but I found it pretty handy to have it in the bundled library, as other libraries can use it as the common interface.  If we can invent the standard library, it may be an option.